### PR TITLE
Feature/condo/sberdoma 111/property units builder

### PR DESCRIPTION
**Pull request by dkoviazin**
Add json schema validator

chess1
chess2
chess3
chess4
chess5
chess6
chess7

### CODE REVIEW

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.spec.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    })
+
+    describe('Unit operations', () => {
+        describe('Add unit', () => {
+            it('unit should be correctly placed to map', () => {
+                const Building = createBuildingMap(10)
+                const jsonMap = Building.getMap()
+                const newUnit = { id: null, floor: jsonMap.sections[3].floors[3].id, section: jsonMap.sections[3].id, label: '1000' }
+                Building.mapAddUnit(newUnit)
+                Building.validate()
```
Comment (pahaz) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.spec.ts:64]: Why we don't use validate during addUnit call? Is it means that we can add unit and get invalid data structure?

Comment (pahaz) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.spec.ts:64]: It's better to protect developers from their mistakes and don't get them chance to have a mistake.

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+    type: string
+    units: BuildingUnit[]
+}
+
+export type BuildingSection = {
+    id: string
+    index?: number
+    name: string
+    type?: string
+    floors?: BuildingFloor[]
```
Comment (pahaz) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:33]: Why it's optional?

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+    name: string
+    type: string
+    units: BuildingUnit[]
+}
+
+export type BuildingSection = {
+    id: string
+    index?: number
+    name: string
+    type?: string
```
Comment (pahaz) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:32]: Why type is optional?

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+import Ajv from 'ajv'
+
+const ajv = new Ajv()
+const validator = ajv.compile(MapSchemaJSON)
+
+type Maybe<T> = T | null
+
+export type BuildingUnit = {
+    id: Maybe<string>
+    type?: string
```
Comment (pahaz) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:14]: Why type is optional?

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+class MapValid {
+    
+    protected map: BuildingMap
+
+    public isMapValid: boolean
+    public validationError: Maybe<string>
+
+    protected autoincrement: number
+
+    constructor (map: BuildingMap) {
```
Comment (pahaz) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:66]: map?

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+}
+
+type IndexLocation = {
+    section: number
+    floor: number
+    unit: number
+}
+
+
+class MapValid {
```
Comment (pahaz) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:57]: Map validator?

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+    }
+
+    private updateFormField () {
+        this.updateMap(this.map)
+    }
+    private updateMap: (map: BuildingMap) => void
+}
+
+
+class MapDebug extends MapEdit {
```
Comment (pahaz) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:472]: Do we really need it?

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+        const unitIndex = this.getUnitIndex(id)
+        if (unitIndex.unit !== -1) {
+            this.map.sections[unitIndex.section].floors[unitIndex.floor].units.splice(unitIndex.unit, 1)
+        }
+        this.selectedUnit = null
+        this.editMode = 'addUnit'
+        this.updateFormField()
+    }
+
+    public mapUpdateUnit (unit: BuildingUnit): void {
```
Comment (pahaz) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:446]: updateUnit, no need to add `map` prefix

File: apps/condo/domains/property/components/panels/Builder/MapJsonSchema.json [hunk 9]
```diff
@@ -0,0 +1,98 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "properties": {
+        "dv": {
+            "type": "number"
```
Comment (pahaz) [inline apps/condo/domains/property/components/panels/Builder/MapJsonSchema.json:5]: Should be equal to 1

File: apps/condo/domains/property/schema/Property.js [hunk 10]
```diff
@@ -75,7 +80,10 @@ const Property = new GQLListSchema('Property', {
                     if (typeof value !== 'object') { return addFieldValidationError(`${JSON_EXPECT_OBJECT_ERROR}${fieldPath}] ${fieldPath} field type error. We expect JSON Object`) }
                     const { dv } = value
                     if (dv === 1) {
-                        // TODO(pahaz): need to checkIt! sections and so on
+                        if (!jsonMapValidator(value)){
```
Comment (pahaz) [inline apps/condo/domains/property/schema/Property.js:83]: We need to add at least two tests for that (one positive and negative)

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 6 above)
Comment (pahaz) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:57]: Or validMap? May be just map ?

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+    
+    private setAutoincrement () {
+        this.autoincrement = Math.max(0, ...this.map.sections
+            .map(section => section.floors
+                .map(floor => floor.units
+                    .map(unit => !isNaN(+unit.id) ? Number(unit.id) : 0)))
+            .flat(2)) + 1
+    }
+
+    private updateStructure (): void {
```
Comment (pahaz) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:85]: It looks like `repairMapStructure`.

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 12]
```diff
@@ -0,0 +1,490 @@
...
+        this.autoincrement = 0
+        this.map.sections.forEach((section, idx) => {
+            section.type = 'section'
+            section.id = String(++this.autoincrement)
+            section.index = idx
+            section.floors.forEach(floor => {
+                floor.type = 'floor'
+                floor.id = String(++this.autoincrement)
```
Comment (pahaz) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:93]: Floor also have index and name. We can rename floor and we also have technical floors

File: apps/condo/domains/property/components/panels/Builder/MapJsonSchema.json [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+                                            "label": {
+                                                "type": "string"
+                                            },
+                                            "type": {
+                                                "type": "string"
+                                            }
+                                        },
+                                        "required": [
+                                            "id",
+                                            "label",
```
Comment (pahaz) [inline apps/condo/domains/property/components/panels/Builder/MapJsonSchema.json:40]: It's better to add name here. And we have a difference between names and label's. Names is hand written but label's is always generated

File: apps/condo/domains/property/components/UnitButton.tsx [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+            ButtonLabel = `…${ButtonLabel.substring(ButtonLabel.length - 2)}`
+        } else {
+            ButtonLabel = `${ButtonLabel.substring(0, 2)}…`
+        }
+        return (
+            <Tooltip placement='topLeft' title={OriginalLabel}>
+                <Button css={css`
+                    ${buttonCss};
+                    ${selected ? selectedCss : ''};
+                    ${selected ? selectedCss : ''};
```
Comment (Dimitreee) [inline apps/condo/domains/property/components/UnitButton.tsx:142]: Duplicates previous string

Comment (Dimitreee) [reply apps/condo/domains/property/components/UnitButton.tsx:142]: can we use classNames for css classNames composition

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.spec.ts (hunk 1 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.spec.ts:64]: I move validation to another section. Now there are 4 more tests to check validator and to check that operations with map and union produce valid json

File: apps/condo/pages/property/index.tsx [hunk 15]
```diff
@@ -32,12 +32,15 @@ import { useTableColumns } from '@condo/domains/property/hooks/useTableColumns'
 import { Property } from '@condo/domains/property/utils/clientSchema'
 
 import debounce from 'lodash/debounce'
+import { isEmpty } from 'lodash'
```
Comment (Dimitreee) [inline apps/condo/pages/property/index.tsx:35]:
```suggestion
import isEmpty from 'lodash/isEmpty'
```

i'm not sure but there is some problems with lodash full import when treeshaking is applyin by webpack in prod

File: apps/condo/domains/property/components/panels/Builder/BuildingPanelEdit.tsx [hunk 16]
```diff
@@ -0,0 +1,468 @@
...
+    BuildingUnit,
+    BuildingSection,
+} from './MapConstructor'
+
+import ScrollContainer from 'react-indiana-drag-scroll'
+
+const { Option } = Select
+
+const INPUT_STYLE = {
+    width: '136px',
```
Comment (Dimitreee) [inline apps/condo/domains/property/components/panels/Builder/BuildingPanelEdit.tsx:26]: 136px?)) looks strange

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 17]
```diff
@@ -0,0 +1,490 @@
...
+        this.isMapValid = this.validate()
+        if (!this.isMapValid) {
+            this.updateStructure() 
+            this.isMapValid = this.validate()        
+        }
+        this.setAutoincrement()
+    }
+    
+    private setAutoincrement () {
+        this.autoincrement = Math.max(0, ...this.map.sections
```
Comment (Dimitreee) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:78]: What will happend when autoincrement is already setted?

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 18]
```diff
@@ -0,0 +1,490 @@
...
+                floor.id = String(++this.autoincrement)
+                floor.units.forEach(unit => {
+                    unit.type = 'unit'
+                    unit.id = String(++this.autoincrement)
+                })
+            })
+        })
+    }
+
+    validateUniqueId (): boolean {
```
Comment (Dimitreee) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:102]: add Typescript class field accessors, `public, private, static`

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 11 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:85]: Fixed

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 5 above)
Comment (Dimitreee) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:66]: add property accessor

```suggestion
    constructor (private/public/protected map: BuildingMap) {
```

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 19]
```diff
@@ -0,0 +1,490 @@
...
+        }
+        return true
+    }
+
+}
+
+
+class MapView extends MapValid {
+
+    constructor (map: Maybe<BuildingMap>) {
```
Comment (Dimitreee) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:147]:
```suggestion
    constructor (private/public/protected map: Maybe<BuildingMap>) {
```

File: apps/condo/domains/property/components/panels/Builder/MapJsonSchema.json (hunk 13 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapJsonSchema.json:40]: Done

File: apps/condo/pages/property/index.tsx (hunk 15 above)
Comment (dkoviazin) [reply apps/condo/pages/property/index.tsx:35]: Done

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 20]
```diff
@@ -0,0 +1,490 @@
...
+        const result = { section: -1, floor: -1, unit: -1 }
+        try {
+            this.map.sections.forEach((section, sectionIdx) => {
+                section.floors.forEach((floor, floorIdx) => {
+                    const unitIdx = floor.units.findIndex(unit => unit.id === unitId)
+                    if (unitIdx !== -1) {
+                        result.section = sectionIdx
+                        result.floor = floorIdx
+                        result.unit = unitIdx
+                        throw new Error('Found')
```
Comment (Dimitreee) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:210]: redurant throw/try/catch

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 17 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:78]: I removed autoincrement from map-JSON structure to object property - so it's only a service field to create the next id.
If we decide to use unique ids for units then autoincrement will be removed

File: apps/condo/domains/property/components/panels/Builder/BuildingPanelEdit.tsx (hunk 16 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/BuildingPanelEdit.tsx:26]: Pixel perfect from design

File: apps/condo/domains/property/schema/Property.js (hunk 10 above)
Comment (dkoviazin) [reply apps/condo/domains/property/schema/Property.js:83]: Done

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 21]
```diff
@@ -0,0 +1,490 @@
...
+    }
+    
+}
+
+
+class MapEdit extends MapView {
+
+    constructor (map: Maybe<BuildingMap>, updateMap: Maybe<(map: BuildingMap) => void>) {
+        super(map)
+        this.updateMap = updateMap
```
Comment (Dimitreee) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:260]:
```suggestion
        this.updateMap = updateMap
```

redurant command, all constructor arguments dirrectly apllying to 'this' with same name

https://www.typescriptlang.org/play?#code/MYGwhgzhAEAKBOB7ADtA3gKGt6xEDsIAXeAV2CMXgApl4BLANzCIFNoAze+YgLmIb4A5gBpodJi3YRWefABN+JesICU6LDi1EAFvQgA6LjyLQAvJ27FN2AL4ZbQA

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 5 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:66]: It's a field name from a property object = my imagination fails at this point. Maybe, use chessUnitsBoard or smth else to make it similar to 'Шахматка'

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 4 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:14]: Done

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 3 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:32]: Done

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 2 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:33]: Done

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 22]
```diff
@@ -0,0 +1,490 @@
...
+    }
+    
+    public mapRemoveSection (id: string): void {
+        const sectionIndex = this.map.sections.findIndex(mapSection => mapSection.id === id)
+        this.map.sections.splice(sectionIndex, 1)
+        this.editMode = 'addSection'
+        this.updateFormField()
+    }
+
+    public mapAddUnit (unit: BuildingUnit): void {
```
Comment (Dimitreee) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:412]: `move verbs on the start of the method/function name's`

i'll comment this one earlier

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 6 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:57]: Renamed

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 7 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:472]: Removed - I use it in tests to have access to JSON. But I think there is no need to protect JSON - as we have validation on save

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 8 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:446]: Done

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 18 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:102]: Done

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 23]
```diff
@@ -0,0 +1,490 @@
...
+            .flat(2)) + 1
+        return result
+    }
+
+    private updateUnitNumbers (unitFrom: BuildingUnit): void {
+        const { id, label } = unitFrom
+        if (isNaN(+label)) {
+            return
+        }
+        let started = false
```
Comment (Dimitreee) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:277]: ??

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 20 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:210]: Done

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts [hunk 24]
```diff
@@ -0,0 +1,490 @@
...
+            this.updateUnitNumbers(unit)
+        }
+        this.editMode = 'editUnit'
+        this.updateFormField()
+    }
+
+    private updateFormField () {
+        this.updateMap(this.map)
+    }
+    private updateMap: (map: BuildingMap) => void
```
Comment (Dimitreee) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:468]: Duplicate type declaration

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 19 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:147]: Done

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 21 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:260]: Done

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 23 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:277]: Looks cool :) For now - replaced
```js
+label
```
to 
```js 
Number(label)
```
Need to change some logic after finding out that Unit can have name

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 24 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:468]: Done

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.ts (hunk 5 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.ts:66]: Done

File: apps/condo/domains/property/components/panels/Builder/MapJsonSchema.json (hunk 9 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapJsonSchema.json:5]: Done

File: apps/condo/domains/property/components/UnitButton.tsx (hunk 14 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/UnitButton.tsx:142]: Skipped "can we use classNames for css classNames composition?" will return to it later
duplicates  - fixed

**Comment by Dimitreee** [comment]
Screenshot 2021-05-03 at 07 01 17

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.spec.ts [hunk 25]
```diff
@@ -0,0 +1,209 @@
...
+const createBuildingMap = (sections: number) => {
+    const PropertyMap = new MapEdit(null, () => null )
+    for (let i = 0; i < sections; i++){
+        PropertyMap.addSection({ ...testSection, name: sectionName(), type: MapTypesList.Section })
+    }
+    return PropertyMap
+}
+
+
+describe('Map constructor', () => {
```
Comment (pahaz) [inline apps/condo/domains/property/components/panels/Builder/MapConstructor.spec.ts:24]: could you please provide structure tests like in BBuilderData? It's hard to understand is it backward compatible or not. Without the state tests it's hard to understand that logic too

**Comment by pahaz** [comment]
If you don't need the BBuilder it's better to remove it.

**Review by Dimitreee — APPROVED** [review]

File: apps/condo/domains/property/components/panels/Builder/MapConstructor.spec.ts (hunk 25 above)
Comment (dkoviazin) [reply apps/condo/domains/property/components/panels/Builder/MapConstructor.spec.ts:28]: 1. Moved old JSONS from BBuilderData to new tests
2. Added new logic to auto repair mechanics (as old JSON misses some key fields like name, dv, type ....) 
3. Added tests to check result structure to match JSON structure
4. Old tests have some checks on old matrix mechanics which I didn't implement so I skip these tests
